mismi-s3/src/Mismi/S3/Commands.hs

```diff
@@ -150,24 +154,41 @@ move source destination' =
   copy source destination' >>
     delete source
 
-upload :: FilePath -> Address -> AWS ()
+upload :: FilePath -> Address -> AWS UploadResult
 upload =
   uploadWithMode Fail
 
-uploadWithMode :: WriteMode -> FilePath -> Address -> AWS ()
-uploadWithMode m f a = do
-  when (m == Fail) . whenM (exists a) . throwM . DestinationAlreadyExists $ a
-  unlessM (liftIO $ doesFileExist f) . throwM . SourceFileMissing $ f
+uploadOrFail :: FilePath -> Address -> AWS ()
+uploadOrFail f a =
+  upload f a >>= liftUploadResult
+
+uploadWithModeOrFail :: WriteMode -> FilePath -> Address -> AWS ()
+uploadWithModeOrFail w f a =
+  uploadWithMode w f a >>= liftUploadResult
+
+liftUploadResult :: UploadResult -> AWS ()
```

Comment:
FWIW This can be more generic if you like: `MonadThrow m => UploadResult -> m ()`